pkg/internal/controller/controller.go

```diff
@@ -91,13 +101,52 @@ type Controller struct {
 	LeaderElected *bool
 }
 
+type clusterDescription struct {
+	cluster.Cluster
+	ctx     context.Context
+	sources []source.Source
+}
+
 // watchDescription contains all the information necessary to start a watch.
 type watchDescription struct {
 	src        source.Source
 	handler    handler.EventHandler
 	predicates []predicate.Predicate
 }
 
+func (w *watchDescription) IsClusterAware() bool {
+	if _, ok := w.src.(cluster.AwareDeepCopy[*internal.Kind]); !ok {
+		if _, ok := w.src.(cluster.AwareDeepCopy[source.Source]); !ok {
+			return false
+		}
+	}
+	if _, ok := w.handler.(cluster.AwareDeepCopy[handler.EventHandler]); !ok {
+		return false
+	}
+	return true
+}
+
+func (w *watchDescription) DeepCopyFor(c cluster.Cluster) *watchDescription {
+	copy := &watchDescription{
+		predicates: w.predicates,
```

Comment:
nit: This shares the predicates, right? Is that fine or is there any way we can copy them?